web-runtime/src/module/memory.rs

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2020 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::cell::Cell;
+use it_traits::{MemoryAccessError, SequentialWriter};
+use it_traits::SequentialReader;
+use wasmer_it::interpreter::wasm;
+use crate::js_log;
+use crate::marine_js::WasmMemory;
+
+pub(super) struct WITMemoryView {
+    module_name: String,
+}
+
+impl WITMemoryView {
+    pub fn new(module_name: String) -> Self {
+        crate::js_log("WITMemoryView::new called");
+
+        Self { module_name }
+    }
+
+    fn check_bounds(
+        &self,
+        offset: usize,
+        size: usize,
+        memory_size: usize,
+    ) -> Result<(), MemoryAccessError> {
+        if offset + size >= memory_size {
+            Err(MemoryAccessError::OutOfBounds {
+                offset,
+                size,
+                memory_size,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub(super) struct JsSequentialReader {
+    offset: Cell<usize>,
+    data: Vec<u8>,
+    memory: WasmMemory,
+    start_offset: usize,
+}
+
+pub(super) struct JsSequentialWriter {
+    offset: usize,
+    data: Vec<Cell<u8>>,
+    current_offset: Cell<usize>,
+    memory: WasmMemory,
+}
+
+impl JsSequentialWriter {
+    pub fn new(offset: usize, size: usize, memory: WasmMemory) -> Self {
+        Self {
+            offset,
+            data: vec![Cell::new(0u8); size],
```

Comment:
It looks like that if you create an instance of `JsSequentialWriter` and then do nothing (don't call any of its method) after dropping it'll rewrite all memory to 0's. Is it intended?

Reply:
You are right about this behaviour and it is not intended. It was not found during testing, because every use of SequentialWriter in interface types exhausts it. I thing its better to read memory there and have a redundant copy than to set up a land mine.

Reply:
Do we really need to make it RAII? What is a purpose of `write_bytes` to write to memory lazy, optimization for a sequence of `write_u8`? If so, maybe we could do a wrapper on top of `JsSequentialWriter` with name `...Guard`, that makes it behaviour obvious.

Reply:
Yes, it is an optimization for sequence of write_u8/write_u32. But these methods are used in interface-types, so the guard should be there. It makes even more sense to drop such behaviour, given that if affects a separate crate in separate repo.